File: Carrier.Swift.SDK/ElastosCarrier/Session/SessionManager.swift
/*
 * Copyright (c) 2018 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import Foundation

@inline(__always) private func TAG() -> String { return "CarrierSessionManager" }

public typealias CarrierSessionRequestHandler = (_ carrier: Carrier,
                                       _ from: String, _ sdp: String) -> Void

/// The class representing carrier session manager.
@objc(ELACarrierSessionManager)
public class CarrierSessionManager: NSObject {

    private static var sessionMgr: CarrierSessionManager?

    private var carrier: Carrier?
    private var handler: CarrierSessionRequestHandler?
    private var didCleanup: Bool

    /// Get a carrier session manager instance.
    ///
    /// This function is convinience way to get instance without interest to
    /// session request from friends.
    ///
    /// - Parameters:
    ///   - carrier: Carrier node instance
    ///   - options: The options to set for carrier session manager
    ///
    /// - Throws: CarrierError
    @objc(initializeSharedInstance:error:)
    public static func initializeSharedInstance(carrier: Carrier) throws {
        if (sessionMgr != nil && sessionMgr!.carrier != carrier) {
            sessionMgr!.cleanup()
        }

        if (sessionMgr == nil) {
            Log.d(TAG(), "Begin to initialize native carrier session manager...")

            var result = ela_session_init(carrier.ccarrier)
            guard result >= 0 else {
                let errno = getErrorCode()
                Log.e(TAG(), "Initialize native session manager error:0x%X", errno)
                throw CarrierError.FromErrorCode(errno: errno)
            }

            Log.d(TAG(), "The native carrier session manager initialized.")

            result = ela_session_set_callback(carrier.ccarrier, nil, nil, nil)
            guard result >= 0 else {
                let errno = getErrorCode()
                Log.e(TAG(), "Set session callback error: 0x%x", errno)
                ela_session_cleanup(carrier.ccarrier)
                throw CarrierError.FromErrorCode(errno: errno)
            }

            sessionMgr = CarrierSessionManager(carrier)
            sessionMgr!.didCleanup = false

            Log.i(TAG(), "Native carrier session manager instance created.");
        }
    }

    /// Get a carrier session manager instance.
    ///
    /// - Parameters:
    ///   - carrier: Carrier node instance
    ///   - options: The options to set for carrier session manager.
    ///   - handler: The handler for carrier session manager to process session
    ///              request from friends.
    ///
    /// - Throws: CarrierError
    @objc(initializeSharedInstance:sessionRequestHandler:error:)
    public static func initializeSharedInstance(carrier: Carrier,
                       sessionRequestHandler handler: @escaping CarrierSessionRequestHandler) throws {
        if (sessionMgr != nil && sessionMgr!.carrier != carrier) {
            sessionMgr!.cleanup()
        }

        if (sessionMgr == nil) {

            Log.d(TAG(), "Begin to initialize native carrier session manager...")

            let sessionManager = CarrierSessionManager(carrier)
            sessionManager.handler = handler

            var result = ela_session_init(carrier.ccarrier)

            guard result >= 0 else {
                let errno = getErrorCode()
                Log.e(TAG(), "Initialize native session manager error: 0x%X", errno)
                throw CarrierError.FromErrorCode(errno: errno)
            }

            let cb: CSessionRequestCallback = { (_, _, cfrom, csdp, _, cctxt) in
                let manager = Unmanaged<CarrierSessionManager>
                        .fromOpaque(cctxt!).takeUnretainedValue()

                let carrier = manager.carrier
                let handler = manager.handler

                let from = String(cString: cfrom!)
                let  sdp = String(cString: csdp!)

                handler!(carrier!, from, sdp)

            }
            let cctxt = Unmanaged.passUnretained(sessionManager).toOpaque()

            result = ela_session_set_callback(carrier.ccarrier, nil, cb, cctxt);
            guard result >= 0 else {
                let errno = getErrorCode()
                Log.e(TAG(), "Set session callback error: 0x%X", errno)
                ela_session_cleanup(carrier.ccarrier)
                throw CarrierError.FromErrorCode(errno: errno)
            }

            Log.d(TAG(), "The native carrier session manager initialized.")

            sessionManager.didCleanup = false
            sessionMgr = sessionManager

            Log.i(TAG(), "Native carrier session manager instance created.");
        }
    }

    /// Get a carrier session manager instance.
    ///
    /// - Returns: The carrier session manager or nil
    public static func sharedInstance() -> CarrierSessionManager? {
        return sessionMgr;
    }

    private init(_ carrier: Carrier) {
        self.carrier = carrier
        self.didCleanup = true
        super.init()
    }

    deinit {
        cleanup()
    }

    ///  Clean up carrier session manager.
    public func cleanup() {

        objc_sync_enter(self)
        if !didCleanup {
            Log.d(TAG(), "Begin clean up native carrier session manager ...")

            ela_session_cleanup(carrier!.ccarrier)
            carrier = nil
            CarrierSessionManager.sessionMgr = nil
            didCleanup = true

            Log.i(TAG(), "Native carrier session managed cleanuped.")
        }
        objc_sync_exit(self)
    }

    /// Create a new session converstation to the specified friend.
    ///
    /// The session object represent a conversation handle to a friend.
    ///
    /// - Parameters:
    ///   - target:    The target id.
    ///
    /// - Returns: The new CarrierSession
    ///
    /// - Throws: CarrierError
    public func createSession(to target: String)
        throws -> CarrierSession {

        let ctmp = target.withCString { (ptr) -> OpaquePointer? in
            return ela_session_new(carrier!.ccarrier, ptr)
        }

        guard ctmp != nil else {
            let errno = getErrorCode()
            Log.e(TAG(), "Open session conversation to \(target) error: 0x%X", errno)
            throw CarrierError.FromErrorCode(errno: errno)
        }

        Log.i(TAG(), "An new session to \(target) created locally.")

        return CarrierSession(ctmp!, target)
    }
}

File: Carrier.Android.SDK/app/src/main/cpp/sessionManager.c
/*
 * Copyright (c) 2018 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include <jni.h>
#include <assert.h>
#include <stdlib.h>
#include <string.h>
#include <ela_carrier.h>
#include <ela_session.h>

#include "log.h"
#include "utils.h"
#include "utilsExt.h"
#include "carrierCookie.h"
#include "sessionUtils.h"

typedef struct CallbackContext {
    JNIEnv* env;
    jclass  clazz;
    jobject carrier;
    jobject handler;
} CallbackContext;

static inline
CallbackContext* getCallbackContext(JNIEnv* env, jobject thiz)
{
    uint64_t ctxt = 0;
    return getLongField(env, thiz, "nativeCookie", &ctxt) ? \
           (CallbackContext*)ctxt : NULL;
}

static
void onSessionRequestCallback(ElaCarrier* carrier, const char* from, const char *bundle,
                              const char* sdp, size_t len, void* context)
{
    CallbackContext* cc = (CallbackContext*)context;
    int needDetach = 0;
    JNIEnv* env;
    jstring jfrom;
    jstring jsdp;

    assert(carrier);
    assert(from);
    // assert(!bundle);
    assert(sdp);

    (void)carrier;
    (void)len;

    env = attachJvm(&needDetach);
    if (!env) {
        logE("Attach JVM error");
        return;
    }

    jfrom = (*env)->NewStringUTF(env, from);
    if (!jfrom) {
        detachJvm(env, needDetach);
        return;
    }

    jsdp = (*env)->NewStringUTF(env, sdp);
    if (!jsdp) {
        (*env)->DeleteLocalRef(env, jfrom);
        detachJvm(env, needDetach);
        return;
    }

    if (!callVoidMethod(env, cc->clazz, cc->handler, "onSessionRequest",
                        "("_W("Carrier;")_J("String;")_J("String;)V"),
                        cc->carrier, jfrom, jsdp)) {
        logE("Can not call method:\n\tvoid onSessionRequest(Carrier, String, String)");
    }

    (*env)->DeleteLocalRef(env, jsdp);
    (*env)->DeleteLocalRef(env, jfrom);

    detachJvm(env, needDetach);
}

static
bool callbackCtxtSet(CallbackContext* hc, JNIEnv* env, jobject jcarrier, jobject jhandler)
{

    jclass lclazz;
    jclass  gclazz;
    jobject gjcarrier;
    jobject gjhandler;

    lclazz = (*env)->GetObjectClass(env, jhandler);
    if (!lclazz) {
        setErrorCode(ELA_GENERAL_ERROR(ELAERR_LANGUAGE_BINDING));
        return false;
    }

    gclazz    = (*env)->NewGlobalRef(env, lclazz);
    gjcarrier = (*env)->NewGlobalRef(env, jcarrier);
    gjhandler = (*env)->NewGlobalRef(env, jhandler);

    if (!gclazz || !gjcarrier || !gjhandler) {
        setErrorCode(ELA_GENERAL_ERROR(ELAERR_OUT_OF_MEMORY));
        goto errorExit;
    }

    hc->env     = NULL;
    hc->clazz   = gclazz;
    hc->carrier = gjcarrier;
    hc->handler = gjhandler;
    return true;

errorExit:
    if (gjhandler) (*env)->DeleteGlobalRef(env, gjhandler);
    if (gjcarrier) (*env)->DeleteGlobalRef(env, gjcarrier);
    if (gclazz)    (*env)->DeleteGlobalRef(env, gclazz);

    return false;
}

static
void callbackCtxtCleanup(CallbackContext* cc, JNIEnv* env)
{
    assert(cc);

    if (cc->clazz)
        (*env)->DeleteGlobalRef(env, cc->clazz);
    if (cc->carrier)
        (*env)->DeleteGlobalRef(env, cc->carrier);
    if (cc->handler)
        (*env)->DeleteGlobalRef(env, cc->handler);

    free(cc);
}

static
jboolean sessionMgrInit(JNIEnv* env, jobject thiz, jobject jcarrier, jobject jhandler)
{
    CallbackContext *hc = NULL;
    ElaCarrier *carrier = NULL;
    int rc;

    assert(jcarrier);

    carrier = getCarrier(env, jcarrier);

    rc = ela_session_init(carrier);
    if (rc < 0) {
        logE("Call ela_session_init API error");
        setErrorCode(ela_get_error());
        return JNI_FALSE;
    }

    if (!jhandler)
        return JNI_TRUE;

    hc = malloc(sizeof(CallbackContext));
    if (!hc) {
        ela_session_cleanup(carrier);
        setErrorCode(ELA_GENERAL_ERROR(ELAERR_OUT_OF_MEMORY));
        return JNI_FALSE;
    }
    memset(hc, 0, sizeof(*hc));

    if (!callbackCtxtSet(hc, env, jcarrier, jhandler)) {
        callbackCtxtCleanup(hc, env);
        ela_session_cleanup(carrier);
        setErrorCode(ELA_GENERAL_ERROR(ELAERR_LANGUAGE_BINDING));
        return JNI_FALSE;
    }

    rc = ela_session_set_callback(carrier, NULL, onSessionRequestCallback, hc);
    if (rc < 0) {
        callbackCtxtCleanup(hc, env);
        ela_session_cleanup(carrier);
        logE("Call ela_session_set_callback API error");
        setErrorCode(ela_get_error());
        return JNI_FALSE;
    }

    setLongField(env, thiz, "nativeCookie", (uint64_t)hc);

    return JNI_TRUE;
}

static
void sessionMgrCleanup(JNIEnv* env, jobject thiz, jobject jcarrier)
{
    CallbackContext *hc = getCallbackContext(env, thiz);

    assert(jcarrier);

    if (hc) {
        callbackCtxtCleanup(hc, env);
        setLongField(env, thiz, "nativeCookie", 0);
    }

    ela_session_cleanup(getCarrier(env, jcarrier));
}

static
jobject createSession(JNIEnv* env, jclass clazz, jobject jcarrier, jstring jto)
{
    const char *to;
    ElaSession *session;
    jobject jsession;

    assert(jcarrier);
    assert(jto);

    to = (*env)->GetStringUTFChars(env, jto, NULL);
    if (!to) {
        setErrorCode(ELA_GENERAL_ERROR(ELAERR_LANGUAGE_BINDING));
        return NULL;
    }

    session = ela_session_new(getCarrier(env, jcarrier), to);
    (*env)->ReleaseStringUTFChars(env, jto, to);
    if (!session) {
        logE("Call ela_session_new API error");
        setErrorCode(ela_get_error());
        return NULL;
    }

    if (!newJavaSession(env, session, jto, &jsession)) {
        ela_session_close(session);
        setErrorCode(ELA_GENERAL_ERROR(ELAERR_LANGUAGE_BINDING));
        return NULL;
    }

    return jsession;
}

static
jint getErrorCode(JNIEnv* env, jclass clazz)
{
    (void)env;
    (void)clazz;

    return _getErrorCode();
}

static const char* gClassName = "org/elastos/carrier/session/Manager";
static JNINativeMethod gMethods[] = {
    {"native_init",      "("_W("Carrier;")_S("ManagerHandler;)Z"),  (void*)sessionMgrInit   },
    {"native_cleanup",   "("_W("Carrier;)V"),                       (void*)sessionMgrCleanup},
    {"create_session",   "("_W("Carrier;")_J("String;)")_S("Session;"),
                                                                    (void*)createSession    },
    {"get_error_code",   "()I",                                     (void*)getErrorCode     },
};

int registerCarrierSessionManagerMethods(JNIEnv* env)
{
    return registerNativeMethods(env, gClassName,
                                 gMethods,
                                 sizeof(gMethods) / sizeof(gMethods[0]));
}

void unregisterCarrierSessionManagerMethods(JNIEnv* env)
{
    jclass clazz = (*env)->FindClass(env, gClassName);
    if (clazz)
        (*env)->UnregisterNatives(env, clazz);
}